### ISSUE TITLE
[Feature Request] Add support for parent chd files

### ISSUE THREAD

**Issue by Sanaki** [libretro/beetle-psx-libretro#587]
Recently I came across the ability to generate chd files with a parent specified, where data present in the parent is referenced rather than being duplicated in the child. For most platforms, the main usage would be in keeping multiple revisions or applying minor hacks/translations. For PS1 though, there's a bit of an extra use case in handling the massive number of multi-disc games the system offers. I tested the process out on Legend of Dragoon for the sake of curiosity, using disc 1 as the parent for the others.
```
222M Legend of Dragoon, The (USA) (Disc 1).chd
252M Legend of Dragoon, The (USA) (Disc 2).chd
164M Legend of Dragoon, The (USA) (Disc 2) (child).chd
299M Legend of Dragoon, The (USA) (Disc 3).chd
212M Legend of Dragoon, The (USA) (Disc 3) (child).chd
354M Legend of Dragoon, The (USA) (Disc 4).chd
266M Legend of Dragoon, The (USA) (Disc 4) (child).chd
```
Grand total space saved on this particular example was 263MB.

So far as I can tell, the parent is only referenced by its chd sha1, so there is a distinct complication in identifying the parent. That's the biggest issue I can see with implementation.

Syntax for generation:
`chdman createcd -i "disc.cue" -op "parent.chd" -o "disc.chd"`

**Comment by bslenul** [comment]
Never heard of that parent/child CHD before, this is pretty cool!

Quick test with Final Fantasy VII:
* Disc 1 (parent): 460MB
* Disc 2: 248MB (clone) instead of 443MB with normal conversion
* Disc 2: 175MB (clone) instead of 400MB

So the whole game is ~883MB with parent/child, instead of ~1.3GB with regular CHD, that's really nice :D

**Comment by Sanaki** [comment]
Another decent one, 382MB saved:
```
336M Tales of Destiny II (USA) (Disc 1).chd
356M Tales of Destiny II (USA) (Disc 2).chd
165M Tales of Destiny II (USA) (Disc 2) (child).chd
393M Tales of Destiny II (USA) (Disc 3).chd
202M Tales of Destiny II (USA) (Disc 3) (child).chd
```

**Comment by i30817** [comment]
This is lovely but it's probably not a good idea to change  the names of the discs from the redump names unless you want the scanner to pretend they don't exist (unless the scanner treats chd special and uses a header name or something)

**Comment by Sanaki** [comment]
The scanner doesn't care about filenames at all, nor does the core. The only time that matters is when a ToC type is used, such as a cue sheet. Unlike those, chd stores that information within the file as metadata. That said, I only named them child to keep them distinct for testing.

<details>
  <summary>Example CHD info for Darkstone (many tracks)</summary><p>

```
chdman - MAME Compressed Hunks of Data (CHD) manager 0.217 (mame0217-423-g05f5625366)
Input file:   Darkstone (USA).chd
File Version: 5
Logical size: 754,972,992 bytes
Hunk Size:    19,584 bytes
Total Hunks:  38,551
Unit Size:    2,448 bytes
Total Units:  308,404
Compression:  cdlz (CD LZMA), cdzl (CD Deflate), cdfl (CD FLAC)
CHD size:     393,044,462 bytes
Ratio:        52.1%
SHA1:         0e8d09008e3be03229e9c7aa4d747225b32286f4
Data SHA1:    d731966609d90e2c99fc646486a2cbd9a26c1c38
Metadata:     Tag='CHT2'  Index=0  Length=93 bytes
              TRACK:1 TYPE:MODE2_RAW SUBTYPE:NONE FRAMES:159088 PREGAP:0 PGTYPE:MODE1 PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=1  Length=90 bytes
              TRACK:2 TYPE:AUDIO SUBTYPE:NONE FRAMES:5989 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=2  Length=90 bytes
              TRACK:3 TYPE:AUDIO SUBTYPE:NONE FRAMES:5070 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=3  Length=90 bytes
              TRACK:4 TYPE:AUDIO SUBTYPE:NONE FRAMES:5077 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=4  Length=90 bytes
              TRACK:5 TYPE:AUDIO SUBTYPE:NONE FRAMES:4738 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=5  Length=91 bytes
              TRACK:6 TYPE:AUDIO SUBTYPE:NONE FRAMES:10426 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=6  Length=91 bytes
              TRACK:7 TYPE:AUDIO SUBTYPE:NONE FRAMES:10025 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=7  Length=91 bytes
              TRACK:8 TYPE:AUDIO SUBTYPE:NONE FRAMES:10556 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=8  Length=90 bytes
              TRACK:9 TYPE:AUDIO SUBTYPE:NONE FRAMES:9826 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=9  Length=92 bytes
              TRACK:10 TYPE:AUDIO SUBTYPE:NONE FRAMES:10742 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=10  Length=91 bytes
              TRACK:11 TYPE:AUDIO SUBTYPE:NONE FRAMES:6700 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=11  Length=91 bytes
              TRACK:12 TYPE:AUDIO SUBTYPE:NONE FRAMES:5519 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=12  Length=91 bytes
              TRACK:13 TYPE:AUDIO SUBTYPE:NONE FRAMES:5356 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=13  Length=91 bytes
              TRACK:14 TYPE:AUDIO SUBTYPE:NONE FRAMES:4718 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=14  Length=91 bytes
              TRACK:15 TYPE:AUDIO SUBTYPE:NONE FRAMES:5879 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=15  Length=91 bytes
              TRACK:16 TYPE:AUDIO SUBTYPE:NONE FRAMES:6129 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=16  Length=92 bytes
              TRACK:17 TYPE:AUDIO SUBTYPE:NONE FRAMES:14273 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=17  Length=91 bytes
              TRACK:18 TYPE:AUDIO SUBTYPE:NONE FRAMES:5553 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=18  Length=92 bytes
              TRACK:19 TYPE:AUDIO SUBTYPE:NONE FRAMES:14371 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.
Metadata:     Tag='CHT2'  Index=19  Length=91 bytes
              TRACK:20 TYPE:AUDIO SUBTYPE:NONE FRAMES:8333 PREGAP:150 PGTYPE:VAUDIO PGSUB:NONE POSTGAP:0.

     Hunks  Percent  Name
----------  -------  ------------------------------------
       341     0.9%  Copy from self                          
    16,940    43.9%  CD LZMA                                 
     3,983    10.3%  CD Deflate                              
    17,287    44.8%  CD FLAC
```
</p></details>

As for hacks, it depends on what your intention is and what the hack is. If you want to keep both the hacked version and the unhacked version and the hack itself isn't invasive (that is, it only modifies a few files rather than -everything-), your best bet would be to use the originals as the parents for the hacks. Disc 1 clean as parent to disc 1 hack, disc 2 clean as parent to disc 2 hack, etc. You'd end up with hacked chds potentially under 1MB that way. If you aren't keeping the originals, you'd just use the same method I did, with disc 1 as parent, and see what you get. I'd expect similar results between hacked and unhacked, since hacks likely would apply the same edits to the same shared files on each disc.

**Comment by i30817** [comment]
While that serial thing is true, it's also not true if you want to use the 'manual' scanner, which is one of the only solutions (with the right dat) to get in playlists hacked game names without them being misidentified as the original game. I haven't tried that yet, mostly because of laziness, but i did try the MAME split set dat (which isn't usually supported) and it created the playlist (albeit without images).

**Comment by Sanaki** [comment]
I don't really work with hacks often on disc-based systems. Other than a translation I don't believe I've used any on PS1. If manual scanning is the way you go... well, you do you. As for the recursing bit, I checked earlier (must not have gotten it written in here), you can only designate one file as a parent, and using a child as a parent will only work with data present in that specific file. If disc 1 has chunks a and b, disc 2 has a and c, and disc 3 has a and d, setting 1 as parent to 2 and 2 as parent to 3 would save nothing on 3.

**Comment by i30817** [comment]
opened this issue; which is where this probably should have been opened because libretro is only a consumer of libchd (if i understand it correctly).

https://github.com/rtissera/libchdr/issues/13

**Comment by Sanaki** [comment]
My understanding (possibly flawed) is that libchdr supports it, but that it needs to be fed both files, which isn't currently set up.

That interpretation is based on this comment: https://github.com/mamedev/mame/issues/2634#issuecomment-327303020

**Comment by i30817** [comment]
Uh, that's a weird way of doing it, i'd expect that the parent file would be part of the format. I suppose doing it that way allows renames or different paths to the files; but i can't say i like it.

Meh then, i'll wait for the dev to close that issue then.

If it's supposed to be something given by a dat, making RA support it is likely to be 'difficult' because it wouldn't be a standard dat (from a dumping group).

CHD has user metadata headers though, that could maybe be used to bake in the info in the files as a post processing step. It's probably *possible* to convince individual core devs to accommodate that idea, or at least more possible than the main RA devs doing anything useful, ever, with the scanner.

**Comment by i30817** [comment]
Though in my opinion, you're better off asking for that feature (storing a 'portable' path to the parent as a alternative to dats) in chdman, if you can manage to convince them.

A extra tool would be annoying and a excuse for people not to adopt your new 'unsupported' idea.

**Comment by Sanaki** [comment]
The chd file format as a whole is a MAME feature. It was designed to work with MAME's setup. The dat being the source of the parent name is only relevant for MAME. The only way this would ever enter the purview of retroarch itself is if the various core developers agreed that playlist entries should directly reference the parent in the entry, similar to how history entries store subsystem entries. Barring that happening, it has nothing to do with retroarch, it's a core issue only.

I did consider the metadata option, but that goes back to restricting file names, which isn't optimal. The best solutions I've come up with are as follows:

- Scan directory for chd matching sha1 of parent. This wouldn't involve actually hashing the files, just having libchdr check the internally stored sha1 value on every chd in the same directory.
  - Pros: filename-agnostic, no metadata to futz with, least possible complications encountered by less technically adept users
  - Cons: Large directories could be slow to scan
- Metadata entry identifying the parent
  - Pros: No scanning required
  - Cons: Demands exact filename and path (relative), average user won't be able to change this field, some cores/emus have been known to choke on unknown metadata entries
- Store children in a folder named the same as the parent
  - Pros: Easy to rename if needed, no scanning woes, easy for the average user
  - Cons: Not completely filename-agnostic, will be annoying for some people needing subdirectories in their pristine collections.
- Subsystem load
  - Pros: Directly provides the necessary files based on user input, no scanning needed
  - Cons: No easy way to add these entries to playlists, may not play nice with retroachievements file hashing
- Parent index file containing known parent SHA1s (stored in system or saves most likely)
  - Pros: Removes the need for filesystem rescans on each game load, would fall back to scan if listed parent isn't found in the listed location. Filename-agnostic due to fallback.
  - Cons: Still will scan once for each newly loaded child, requires an external file stored for saving these results.

I understand you have gripes with the way these projects are run, but this isn't the place for that, and it doesn't help the problem at hand. This issue is only for hashing out how to handle -this- issue.

**Comment by i30817** [comment]
You're right on all counts, i overstepped. And forgot about the internal sha1 too! I thought the parent sha1 wouldn't be stored if the path wasn't, glad to see it's not the case.

**Comment by Sanaki** [comment]
If you check the expandable info readout for Darkstone in https://github.com/libretro/beetle-psx-libretro/issues/587#issuecomment-576434339, the SHA1 listed there (not the Data SHA1) is the same one stored as the Parent SHA1 in the child. Using that instead of directly hashing would massively speed up a scan. Without that, pretty sure it wouldn't even be an option worth consideration.

**Comment by i30817** [comment]
From your alternatives, i prefer the first, but only because i put in a game per dir.

However, the question becomes more interesting iff you think about game hacks. If you think of that, you probably want to put them in a separate subdir. But that's no good! You'd have to put the 'main' game on the subdir and the 'hacks' on the parent, which is unnatural.

So i'm thinking that the second or third are more 'natural'. And if this idea gets implemented by other cores, you probably want something everyone can agree on and live with as a 'standard' to not confuse the users further.

**Comment by Sanaki** [comment]
The second is basically an extension of the first. I'm sure there's a way to make it reasonable to use a custom directory structure without needing to manually edit the index file, but I'll need to think on that. I overlooked that desire, to be honest.

**Comment by i30817** [comment]
That structure also isn't very friendly to multiple unrelated children. Imagine two different hacks with different readmes. Of course you can do it with yet another subdir tree, but i don't like it.

Also this is actually a reason to make this part of the scanner. The scanner would treat chd files expecially and build a file (as in alternative 2) once, with all of the data sha1 and paths, and then any core could read that file for the paths, in whatever configuration they were without either user intervention or multiple path crawling.

**Comment by Sanaki** [comment]
Yeah, retroarch playlists are json. Unfortunately, this feature has fairly limited appeal for other systems, so for now I feel like this needs to stay in core until such a time as it's proven expanding to a retroarch-supported feature would be beneficial and not out of scope for the project's goals. I'd love to find out I'm wrong, and that there's a way to implement it without breaking the current standard. When I said scan in the bulleted options above, my thought was for the core to handle that scanning directly, not retroarch.

There's always the option of adding parent handling to the core's options, or as a subsystem maybe (rather than selecting exact files, you provide it an entire directory to scan for children and their requested parents). That would handle the custom directory structure option, since it would recurse subdirectories. We'd need to hear from the core's devs on the matter to know if that solution is viable though.

**Comment by i30817** [comment]
Won't be limited for long imo. CHD support will continue to spread just because it's the MAME format.

**Comment by negativeExponent** [comment]
why would an issue about scanning CHD hinder features like adding ability to play using CHD files? sound like a hate post to be since both issue is unrelated...

**Comment by i30817** [comment]
Because the chd feature needs to find the chd. It's related because the best way (as far i can tell - you might disagree and if so, write it out) to find that file is to store the 'data sha1' of each chd scanned 'somewhere' that the interested cores can find.

This would allow any files to be in any scanned directory, instead of depending on 'conventions' and it would prevent the need for the cores to crawl the filesystem.

**Comment by rz5** [comment]
I'm trying to work on this feature and I am stepping through the CHD open functions to understand what's going on. I'm seeing some problems re

@Zapeth @rtissera - Pinging you both hoping you could provide clarifications on this.

See the following:

https://github.com/libretro/beetle-psx-libretro/blob/9c64fe026e4f9cab88f7f845fdb1bee0abe4df86/deps/libchdr/chd.c#L1338-L1341

https://github.com/libretro/beetle-psx-libretro/blob/9c64fe026e4f9cab88f7f845fdb1bee0abe4df86/deps/libchdr/chd.c#L1915-L1918

In these snippets, libchdr assume all CHD header versions have a 32-bit wide `flags` field on byte 16. This seems to be wrong based on these code comments:

https://github.com/libretro/beetle-psx-libretro/blob/9c64fe026e4f9cab88f7f845fdb1bee0abe4df86/deps/libchdr/chd.h#L130-L148

For the purposes of this feature request, I would need to change how chd.c detects if a file needs a parent, something like this:

```C
chd_error chd_open_file(...)
{
(...)

    /* if we need a parent, make sure we have one */
    if (parent == NULL)
    {
        /* Checks for V4 and below */
        if (newchd->header.version < 5 && (newchd->header.flags & CHDFLAGS_HAS_PARENT))
            EARLY_EXIT(err = CHDERR_REQUIRES_PARENT);
        
        /* Checks for V5 */
        else if (newchd->header.version >= 5 && memcmp(nullsha1, newchd->header.parentsha1, sizeof(newchd->header.parentsha1)) != 0)
            EARLY_EXIT(err = CHDERR_REQUIRES_PARENT);
    }
    
(...)
}
```

After this is addressed, I would run into another problem. This libretro core accepts 1 file as the content. If the file is a CHD clone, the way I'd search for its parent file is to first get the clone's header info.
But I think the libchdr API doesn't allow me to open a clone file without specifying its parent first...

**Comment by rtissera** [comment]
Had a quick look at this.
Internally, it should be somehow easy to add parent / child support to libchdr.
It does need a couple of fixes and implementing missing stuff, but totally doable.

**Comment by rz5** [comment]
Thanks for the feedback. I thought libchdr was ready for parent/clone usage, so I worked on the easy fixes/behavior. But even after successfully opening a clone -  i.e. correctly opening the parent chd and passing its handle while opening the clone - the emulator just hung on the Sony logo.

So if there was doubt about libchdr being able to handle it, that settled it for me.

The work I did is here: https://github.com/libretro/beetle-psx-libretro/commits/r5/chd-clone-support

**Comment by Sanaki** [comment]
Clone support is now included in libchdr. In theory it should now be fully possible to implement.

**Comment by LibretroAdmin** [comment]
OK, so this got implemented
